### PR TITLE
Buffer writes in mem to DB for message events

### PR DESCRIPTION
**Pull request by elimelt**
Instead of writing on every update, buffer them in memory using `MessageBuffer`. This decreases load on DB, and hopefully also increases the RPS we can handle in the `POST /engagement/message` endpoint

### CODE REVIEW

File: server/engagement/buffer.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        try:
+            self._flush_to_db()
+        except Exception as e:
+            logger.exception(f"flush error: {e}")
+
+    def _flush_to_db(self) -> None:
+        if not self._buffer:
+            return
+
+        with self._lock:
```
Comment (bhavik-soni) [inline server/engagement/buffer.py:61]: Should this lock be moved to the top of flush_to_db?

Comment (elimelt) [reply server/engagement/buffer.py:61]: how come? sorry my mind is feeble

Comment (bhavik-soni) [reply server/engagement/buffer.py:61]: I might be interpreting this wrong, but isn't the lock for self._buffer? If so, then don't we want to protect the use of it just above the lock

Comment (elimelt) [reply server/engagement/buffer.py:61]: the intended logic is to make flushes atomic, so we lock and then create a local copy of the buffer `messages` and clear the actual buffer. if the db transaction fails, we add all the updates back to the original buffer. this way the lock doesn't need to be held the entire time we're flushing.

not sure if this is legit though

Comment (bhavik-soni) [reply server/engagement/buffer.py:61]: That makes sense. Worst case it acquires lock with empty buffer and exits anyways

**Review by bhavik-soni — APPROVED** [review]

**Pull request merged**